gamemodes/terrortown/entities/weapons/weapon_tttbase.lua

```diff
@@ -605,111 +568,148 @@ function SWEP:Initialize()
 	end
 end
 
-function SWEP:CalcViewModel()
-	if not CLIENT or not IsFirstTimePredicted() then return end
+if SERVER then
+	local SF_WEAPON_START_CONSTRAINED = 1
 
-	self.bIron = self:GetIronsights()
-	self.fIronTime = self:GetIronsightsTime()
-	self.fCurrentTime = CurTime()
-	self.fCurrentSysTime = SysTime()
-end
+	---
+	-- Picked up by player. Transfer of stored ammo and such.
+	-- @param Player newowner
+	function SWEP:Equip(newowner)
+		if self:IsOnFire() then
+			self:Extinguish()
+		end
 
--- Note that if you override Think in your SWEP, you should call
--- BaseClass.Think(self) so as not to break ironsights
-function SWEP:Think()
-	self:CalcViewModel()
-end
+		self.fingerprints = self.fingerprints or {}
 
----
--- @return boolean
-function SWEP:DyingShot()
-	local fired = false
+		if not table.HasValue(self.fingerprints, newowner) then
+			self.fingerprints[#self.fingerprints + 1] = newowner
+		end
+
+		if self:HasSpawnFlags(SF_WEAPON_START_CONSTRAINED) then
+			-- If this weapon started constrained, unset that spawnflag, or the
+			-- weapon will be re-constrained and float
+			local flags = self:GetSpawnFlags()
+			local newflags = bit.band(flags, bit.bnot(SF_WEAPON_START_CONSTRAINED))
+
+			self:SetKeyValue("spawnflags", newflags)
+		end
+
+		if IsValid(newowner) and self.StoredAmmo > 0 and self.Primary.Ammo ~= "none" then
+			local ammo = newowner:GetAmmoCount(self.Primary.Ammo)
+			local given = math.min(self.StoredAmmo, self.Primary.ClipMax - ammo)
+
+			newowner:GiveAmmo(given, self.Primary.Ammo)
+
+			self.StoredAmmo = 0
+		end
+	end
+
+	---
+	-- @return boolean
+	function SWEP:DyingShot()
+		if not self:GetIronsights() then
+			return false
+		end
 
-	if self:GetIronsights() then
 		self:SetIronsights(false)
 
 		if self:GetNextPrimaryFire() > CurTime() then
-			return fired
+			return false
 		end
 
 		-- Owner should still be alive here
-		if IsValid(self:GetOwner()) then
-			local punch = self.Primary.Recoil or 5
+		if not IsValid(self:GetOwner()) then
+			return false
+		end
 
-			-- Punch view to disorient aim before firing dying shot
-			local eyeang = self:GetOwner():EyeAngles()
-			eyeang.pitch = eyeang.pitch - math.Rand(-punch, punch)
-			eyeang.yaw = eyeang.yaw - math.Rand(-punch, punch)
+		local punch = self.Primary.Recoil or 5
 
-			self:GetOwner():SetEyeAngles(eyeang)
+		-- Punch view to disorient aim before firing dying shot
+		local eyeang = self:GetOwner():EyeAngles()
+		eyeang.pitch = eyeang.pitch - math.Rand(-punch, punch)
+		eyeang.yaw = eyeang.yaw - math.Rand(-punch, punch)
 
-			MsgN(self:GetOwner():Nick() .. " fired his DYING SHOT")
+		self:GetOwner():SetEyeAngles(eyeang)
 
-			self:GetOwner().dying_wep = self
+		MsgN(self:GetOwner():Nick() .. " fired his DYING SHOT")
 
-			self:PrimaryAttack(true)
+		self:GetOwner().dying_wep = self
 
-			fired = true
-		end
+		self:PrimaryAttack(true)
+
+		return true
+	end
+else -- CLIENT
+	function SWEP:CalcViewModel()
```

Comment:
This also should be shared, since it was shared before and is used in a predicted and shared function.